### PR TITLE
Improve split blending logic for Vulkan

### LINKED ISSUE #73677: Distance Directional Shadows Pixelize as you get further away
### Godot version

4.0 rc 1

### System information

Windows 11, Vulkan, GTX 1050 Ti 528.24

### Issue description

Expected Results: Large Distance Directional Shadows start to shrink-fade as you get further from them due to the lower resolution of distant Shadow Splits. This is common Practice in Many Games, Even AAA games such as Shadow of the Tomb Raider for Directional Shadows

Wrong Results: Large Distance Directional Shadows Pixelize as you get further away from them as the shadow blur reduces ultimately resulting in hard shadows in distant splits that look significantly worse compared to simply slowly shrink-fading the shadows as they get less resolution.

This Issue prevents Longer distance Shadows from being usable unless you are intentionally going with Hard shadows which don't have Shadow blur by definition.

While this Issue is also present with Blending Splits off, it is intended behavior there, as with No blending you get Massive seams between two splits of non matching shadow blur radiuses that look significantly worse compared to a little pixelization, especially considering the fact that blending splits is not very useful for smaller distance directional shadows which benefit the most from the change as they are now cheaper to get at good quality

As such https://github.com/godotengine/godot/pull/48776 Should be modified to not work for Blending Splits on, as the usage for it is typically for larger more demanding scenes where a little cost from blending doesn't matter and where the pr causes issues, whereas Blending Splits off should keep current behavior as it provides good quality for smaller scenes where there's no point in forcing a more expensive approach

Expected Results in Alpha 1:

https://user-images.githubusercontent.com/50302479/220336549-df423292-c977-46bb-9d1c-ecd66d5c222a.mp4

Wrong Results in RC 1:

https://user-images.githubusercontent.com/50302479/220336724-045e99ef-c005-4be0-855d-33579fc22aec.mp4

### Steps to reproduce

Enable Blending, and set the Split settings to similar values as in the videos (Higher distance shadows, with significantly lower resolution in the distance compared to close up)

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by mrjustaguy**
Resolves: https://github.com/godotengine/godot/issues/73677

Implemented for Forward+ and Mobile Renderers.
Restored Old Blur Logic when using blending as it significantly improves the visual presentation compared to blending with the Current Blur Logic, and allows for Directional Shadows over Massive distances at good quality when blending is enabled.

The Blending has been hardcoded to start at 10% away from the end of a given split for simplicity's sake, as for most scenarios it works best and for other values you would typically need to be tweaking it on a per split basis along side having per split blur controls for best results, which would be quite complicated both to implement and for most users to properly use

Blending in action
image

Blending disabled
image

### CODE REVIEW

File: servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl [hunk 1]
```diff
@@ -1765,7 +1765,7 @@ void fragment_shader(in SceneData scene_data) {
 
 					pssm_coord /= pssm_coord.w;
 
-					shadow = sample_directional_pcf_shadow(directional_shadow_atlas, scene_data.directional_shadow_pixel_size * directional_lights.data[i].soft_shadow_scale * blur_factor, pssm_coord);
+					shadow = sample_directional_pcf_shadow(directional_shadow_atlas, scene_data.directional_shadow_pixel_size * directional_lights.data[i].soft_shadow_scale * (blur_factor+(1-blur_factor)*float(directional_lights.data[i].blend_splits)), pssm_coord);
```
Comment (akien-mga) [inline servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:1768]: Please follow the surrounding code style (spaces around operators).

You can run clang-format on the .glsl files to apply it automatically.

Comment (mrjustaguy) [reply servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl]: I've added the missing spaces

File: servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl [hunk 2]
```diff
@@ -1799,7 +1799,7 @@ void fragment_shader(in SceneData scene_data) {
 
 						pssm_coord /= pssm_coord.w;
 
-						float shadow2 = sample_directional_pcf_shadow(directional_shadow_atlas, scene_data.directional_shadow_pixel_size * directional_lights.data[i].soft_shadow_scale * blur_factor2, pssm_coord);
+						float shadow2 = sample_directional_pcf_shadow(directional_shadow_atlas, scene_data.directional_shadow_pixel_size * directional_lights.data[i].soft_shadow_scale * (blur_factor2 + ( 1 - blur_factor2) * float(directional_lights.data[i].blend_splits)), pssm_coord);
```
Comment (akien-mga) [inline servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:1802]:
```suggestion
						float shadow2 = sample_directional_pcf_shadow(directional_shadow_atlas, scene_data.directional_shadow_pixel_size * directional_lights.data[i].soft_shadow_scale * (blur_factor2 + (1 - blur_factor2) * float(directional_lights.data[i].blend_splits)), pssm_coord);
```

File: servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl [hunk 3]
```diff
@@ -1765,7 +1765,7 @@ void fragment_shader(in SceneData scene_data) {
 
 					pssm_coord /= pssm_coord.w;
 
-					shadow = sample_directional_pcf_shadow(directional_shadow_atlas, scene_data.directional_shadow_pixel_size * directional_lights.data[i].soft_shadow_scale * blur_factor, pssm_coord);
+					shadow = sample_directional_pcf_shadow(directional_shadow_atlas, scene_data.directional_shadow_pixel_size * directional_lights.data[i].soft_shadow_scale * (blur_factor + (1 - blur_factor) * float(directional_lights.data[i].blend_splits)), pssm_coord);
```
Comment (akien-mga) [inline servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:1768]: You should probably use `1.0 - blur_factor` and not `1`. GPUs need types to be explicit more often that not.

**Comment by akien-mga** [comment]
After doing changes, could you squash the commits? See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) for instructions.

**Comment by mrjustaguy** [comment]
There

**Comment by Calinou** [comment]
I believe the C++ side of the culling code should be changed as well, so that splits don't need to include as many shadows.

Start from here: https://github.com/godotengine/godot/issues/82659

**Comment by mrjustaguy** [comment]
I'm not really good with C++, and am even less familiar with the culling logic, so it'd be better if someone else were to figure that out, to claw some free performance with Split Blending that was enabled, as it's not like this PR is negatively affecting performance given the fact it's rendering the same shadow maps as before when Blending is enabled.

Given the already fairly small difference Blending On vs Off has, I'm not sure how much of a priority that is right now anyhow, as there are bigger fish to fry

**Comment by mrjustaguy** [comment]
Yea this PR is ready for review, my attempts at playing with the culling code have been as expected unfruitful.

However, Not managing to improve the culling code doesn't Mean it was all for nothing! While scrolling around the code my gut picked up a very tiny error that needed to be fixed in order to fix the very long standing issue with 2 Split Stabilization (https://github.com/godotengine/godot/issues/59232).. So I guess that's.. something 🤣

**Review by clayjohn — APPROVED** [review]
Seems fine to me. GLES3 works a bit differently so its okay to leave out. I would like to improve the culling ASAP as this blending logic should allow us to cull may more objects (plus with lawnjelly's recent culling work we should be able to make shadows much faster)

**Comment by Calinou** [comment]
> GLES3 works a bit differently so its okay to leave out.

There is an OpenGL variant of this PR: https://github.com/godotengine/godot/pull/83253

**Review by Calinou — COMMENTED** [review]
Tested locally, it works as expected. I still think this makes shadows too blurry at a distance, and maybe we should find a halfway compromise (or add a property to adjust how much distant shadows should be blurred).

However, it seems this PR doesn't improve performance in a meaningful way, which surprises me. Worse, this logic noticeably *decreases* performance when using the Ultra shadow filter quality setting, likely as a result of the increased blur in the distance.

Tested on a RTX 4090 in 3840×2160. DirectionalLight3D shadow blur property is set to `1.5` on all screenshots. See top-left corner of each screenshot for FPS.

**Testing project:** [test_lightmap_preview_bake_4.x_3.zip](https://github.com/godotengine/godot/files/13187553/test_lightmap_preview_bake_4.x_3.zip)

### Soft Low

Blend Splits Disabled | Enabled, Before | Enabled, After *(this PR)*
-|-|-
Screenshot_20231027_114641 webp | Screenshot_20231027_114737 webp | Screenshot_20231027_114703 webp

### Soft Ultra

Blend Splits Disabled | Enabled, Before | Enabled, After *(this PR)*
-|-|-
Screenshot_20231027_114841 webp | Screenshot_20231027_114859 webp | Screenshot_20231027_114907 webp

**Comment by mrjustaguy** [comment]
We likely need to rework in general how we pass the Blur to the GPU, to allow for adding more fine tune controls, and figure out a good UI for that, but that is true of a lot of the Directional Shadow Controls. See https://github.com/godotengine/godot-proposals/issues/8197 for an overview of changes to consider, though this is for a future undertaking, by someone with a bit more experience with handling the UI side of these things..

**Comment by josefkaragoli** [comment]
I agree with the final cascade perhaps being a bit too blurry, but other than that I think this PR greatly improves the split blending logic and now it makes sense to enable Blend Splits because it visibly improves the appearance of shadows.

**Comment by mrjustaguy** [comment]
Shadow blur and Blending distances should likely get exposed to the user in the future in some form, but that's outside the scope of this PR, and my capabilities to expose

**Comment by YuriSizov** [comment]
Thanks!

**Pull request merged**